python/lsst/pipe/tasks/calibrate.py

```diff
@@ -615,27 +621,29 @@ def run(self, exposure, exposureIdInfo=None, background=None,
             matchMeta = astromRes.matchMeta
             if exposure.getWcs() is None:
                 if self.config.requireAstrometry:
-                    raise RuntimeError(f"WCS fit failed for {exposureIdInfo.expId} and requireAstrometry "
+                    raise RuntimeError(f"WCS fit failed for {idGenerator} and requireAstrometry "
```

Comment:
Yeah, I agree that it's sort of weird to combine the roles of "ID generator" and "data ID formatting" abstraction, especially in a class that's 95% about the former only, but I think it's actually quite convenient in a way that passing another argument for the latter would not be.  The real advantage is that in both cases `IdGenerator` abstracts over whether there really is a data ID or not, so the `run` method doesn't have to care.

Maybe what we need is to put the `__str__` implementation in a better-named method?  I liked `__str__` because it meant we could use the deferred-stringification stuff in `logging` naturally, but that's not going to be a meaningful optimization.

Reply:
I think either we can clearly document the `IdGenerator`s to say that they are used as standins for dataIds in logging in task `run` methods and keep the `str` behavior as-is, or add a `print_data_id` method (or something like that) and then we'd be able to make `str` more relevant to the idgenerator itself.

Reply:
I've done that documenting in the `meas_base` PR now (it's at the top of the `IdGenerator` class docs.